x-pack/plugins/rollup/server/collectors/register.ts

```diff
@@ -136,33 +136,33 @@ async function fetchRollupVisualizations(
   let rollupVisualizations = 0;
   let rollupVisualizationsFromSavedSearches = 0;
 
-  visualizations.forEach((visualization) => {
+  visualizations.forEach((visualizationData) => {
     const {
-      _source: {
-        visualization: {
-          savedSearchRefName,
-          kibanaSavedObjectMeta: { searchSourceJSON },
-        },
-        references = [] as any[],
-      },
-    } = visualization;
+      _source: { visualization, references = [] as any[] },
+    } = visualizationData;
 
-    const searchSource = JSON.parse(searchSourceJSON);
-
-    if (savedSearchRefName) {
-      // This visualization depends upon a saved search.
-      const savedSearch = references.find((ref) => ref.name === savedSearchRefName);
-      if (rollupSavedSearchesToFlagMap[savedSearch.id]) {
-        rollupVisualizations++;
-        rollupVisualizationsFromSavedSearches++;
-      }
-    } else {
-      // This visualization depends upon an index pattern.
-      if (rollupIndexPatternToFlagMap[searchSource.index]) {
-        rollupVisualizations++;
+    if (visualization) {
+      const {
+        savedSearchRefName,
+        kibanaSavedObjectMeta: { searchSourceJSON },
```

Comment:
The failure I stumbled over was related to a malformed object, so destructuring fails. I think these changes indeed fix the specific error I encountered, but it's still possible for this to fail with a similar message (if there is a `visualization` object, but no `kibanaSavedObjectMeta`).

A lot of migration code tries to get that right by checking whether expected paths exist before actually accessing them (e.g. https://github.com/elastic/kibana/blob/d7b830a6bfba6d98206779630fd101cf3f1a0ecf/src/plugins/visualizations/server/saved_objects/visualization_migrations.ts#L239 ).